Retreat now removes tactical AI commands

Arch-Shaman opened the pull request:
Fixes #4411.

Retreat AI would keep the internally generated tactical AI commands in the queue when inserting the move command. This lead to units moving back to whereever the unit AI wanted to move tactically before reengaging fight command. Note that this should not remove user created move commands.

sprunk commented:
A potential worry: a unit with an internal move order will go back to where it started the retreat from (circle on this pic) and then continue on the path the player initially had in mind, but with the internal command removed it will go from the retreat point straight to the destination (via gigaporc on this pic).
![image](https://user-images.githubusercontent.com/2573076/119583744-873d1180-bdc7-11eb-9bcc-668b440ce49f.png)

sprunk commented:
Another thing: it would be good to check for other types of lingering internal commands (Fight spawned from Patrol? Attack spawned from some engine idle behaviour?) and handle them as well.

Arch-Shaman commented:
Patch demo can be found here: http://zero-k.info/Battles/Detail/1123039 (there's some moments when the retreat behavior kind of screws itself over due to pathing into enemy territory during retreat, but otherwise it felt pretty nice)

GoogleFrog commented:
Why not GG with tactical AI for the removal rather than guessing the format of its commands? It would be pretty easy to make a GG that tells tactical AI to cancel the commands it has given, since it remembers where they are and cancels them during normal function.

Is the intention here to remove all internal move commands (which can come from other places) or to only hit those added by tactical AI?

Arch-Shaman commented:
Mm that would work. The intention is for all internal move commands to be cleared. Moving units is dangerous IMO. Move command units do not fight back if they encounter enemy units. They'll happily run into Jacks or other things. In our test version we also removed the idle return move and replaced it with a fight command.

GoogleFrog commented on LuaRules/Gadgets/cmd_retreat.lua:
```diff
@@ -206,6 +206,19 @@ GG.Retreat_ToggleHaven = ToggleHaven
 ----------------------------
 
+local function FixQueue(unitID) -- this gets rid of internally generated commands by tactical ai and keeps user commands.
+	local queue = spGetCommandQueue(unitID, 3)
```
The 3 here seems fragile.
